### PR TITLE
Refactor NNAPI EP OpBuilder/OpSupportChecker structure

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
### Description

As title

-Split long OpBuilder and OpSupportChecker files into individual operator files.

-Add OpBuilder/SupportChecker registry factories.

-Combine the functionality of op_builder and op_support_checker into one op_builder. 
### Motivation and Context

The NNAPI OPBuilder was splitted into OPBuilder (For EP::Compile) and OPSupportChecker (for EP::GetCapability)
At the time it was reasonable choice, but OPBuilder/OPSupportChecker share some logic and has to use addition helper.

Clean up now to make NNAPI OPBuilder/OPSupportChecker into single OPBuilder (similar to what CoreML EP has)

### CODE REVIEW

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc [hunk 1]
```diff
@@ -267,6 +268,39 @@ void GetFlattenOutputShape(const NodeUnit& node_unit, const Shape& input_shape,
   dim_2 = std::accumulate(input_shape.cbegin() + axis, input_shape.cend(), 1, std::multiplies<int32_t>());
 }
 
+// Get the bias size (C) of Gemm op
+// ANEURALNETWORKS_FULLY_CONNECTED only supports 1d bias
+// Will test if C of Gemm can be squeezed and return the 1d vector size after squeeze
+bool GetBiasSize(const Shape& c_shape, int32_t android_feature_level, uint32_t& size) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc:274]: nit: GetBiasSize -> GetGemmBiasSize.  

Function name should clarify it's for Gemm now that it's located in a general purpose helpers class.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.h [hunk 2]
```diff
@@ -40,13 +56,170 @@ Status AddNnapiSplit(ModelBuilder& model_builder,
...
+// the initializer tensor B will be converted from int8 to uint8 by flip each byte by XOR 0x80
+// byte ^ 0x80 == byte + 128
+Status AddInitializerTransposed(ModelBuilder& model_builder,
+                                const OperandType& source_operand_type,
+                                const std::string& name,
+                                bool is_per_tensor_u8s8);
+
+Status ComputeConvPads(const Shape& input_dimen,
+                       const uint32_t weight_size_y, const uint32_t weight_size_x,
+                       const std::vector<int32_t>& onnx_pads, const std::vector<int32_t>& onnx_strides, const std::vector<int32_t>& onnx_dilations,
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.h:103]: nit: line length. There are a few other cpplint warnings to look at.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    // There is no NNAPI support for Android API level 26-
+    return ANEURALNETWORKS_FEATURE_LEVEL_1;
+  }
+
+  virtual bool HasSupportedInputOutputsImpl(
+      const InitializedTensorSet& initializers, const NodeUnit& node_unit,
+      const OpSupportCheckParams& params) const;
+
+  virtual int GetMinSupportedOpSet(const NodeUnit& /* node_unit */) const { return 1; }
+  virtual int GetMaxSupportedOpSet(const NodeUnit& /* node_unit */) const { return 15; }
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.h:76]: Why is this 15? ONNX is up to opset 18 (not officially supported in this ORT release though) and I thought we supported opset 17 currently.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.cc [hunk 4]
```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.h"
+
+namespace onnxruntime {
+namespace nnapi {
+
+bool HasExternalInitializer(const InitializedTensorSet& initializers, const NodeUnit& node_unit) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.cc:9]: Use anonymous namespace if this is a file local helper

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.cc [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+  const auto is_ext_initializer =
+      [&](const NodeArg& node_arg) {
+        const auto& input_name(node_arg.Name());
+        if (!Contains(initializers, input_name))
+          return false;
+
+        const auto& tensor = *initializers.at(input_name);
+        if (tensor.has_data_location() &&
+            tensor.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.cc:18]: Side note: at some point we need to update this to support using initializers directly from the ort format model by checking for kTensorProtoMemoryAddressTag

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc:274]: Although if it's specific to Gemm does it need to be here or could it be in gemm_op_builder.cc?

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.h (hunk 3 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.h]: sorry - accidentally updated with the old version.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc (hunk 1 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc]: moved to gemm_op_builder.cc as a local helper.

also did similar updates for other helper methods only used inside individual op_builder, such as flatten.

**Review by skottmckay — APPROVED** [review]
:shipit:

**Pull request merged**